man/tuxcast.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH tuxcast 1 "FEBRUARY 2010" 0.4 "tuxcast 0.4"
.SH NAME
tuxcast \- The Linux Podcast Client
.SH SYNOPSIS
.B tuxcast 
[\fIOPTIONS\fR]
.SH DESCRIPTION
Tuxcast is a minimal, portable podcatcher written in C++ with 
an interactive command line interface. 

Options and podcast feeds must be configured with
.B tuxcast-config
to make it functional. Then episodes can be downloaded
regularly using the \fB-c\fR option. When feeds are checked, all
undownloaded episodes are gotten unless an option specifies
otherwise.

.SH OPTIONS
.TP 
\fB\-c\fR
Check all configured feeds, print status to screen
.TP 
\fB\-s\fR
Check all configured feeds, very verbose output or graphical mode if
\fBtuxstatus\fR is installed (see GRAPHICAL MODE)
.TP 
\fB\-C name\fR
Check only the specified configured feed
.TP 
\fB\-u [-e num]\fR
Download only the latest episode from all configured feeds or the
latest \fInum\fR episodes using \fB-e\fR
.TP 
\fB\-U name [-e num]\fR
Download only the latest episode from the specified configured feed or the
latest \fInum\fR episodes using \fB-e\fR
.TP
\fB\-q name | URL\fR
Query the given named feed or URL for the available episodes, printing
the episode file names and sizes in bytes
.TP
\fB\-f\fR
Forget about older downloaded episodes (that are no longer in the feeds)
.TP
\fB\-v\fR
Show version, licensing and build configuration information
.TP
\fB\-h\fR
Show a help message

.SH GRAPHICAL MODE
Tuxcast can automatically detect and use the tuxstatus graphical front
end for download status using the \fB-s\fR option. It is enabled if
tuxstatus is installed and the standard output is not going to a
pipe. In this case, the following two commands are identical.
.IP
.B tuxcast -s
.br
.B tuxcast -s 2>/dev/null | tuxstatus -d
.PP
If either of these conditions are not true the original behavior
of very verbose output is used.

.SH FILES
.TP
$HOME/.tuxcast
Directory containing configuration data.
.TP
$HOME/.tuxcast/cache
Directory containing cached feed files.
.TP
$HOME/.tuxcast/config.xml
Main configuration file.
.TP
$HOME/.tuxcast/files.xml
List of episodes downloaded or skipped.
.SH BUGS
There are no known issues in this version of tuxcast.
.SH AUTHOR
Tuxcast is written by David 'Cillian' Turner assistance from Roger 'Oojah' Light, Greatred, and Mathew 'kookjr' Cucuzella.  Bittorrent support is based on the simple_client program for libtorrent, written by Jari 'Rakshasa' Sundell.

These manpages were written by cypher_ of Linux User Group Radio channel (#lugradio) on the freenode network.
.SH "SEE ALSO"
.BR tuxcast-config(1),
.BR tuxstatus(1)

.BR docs
subdirectory in the source distribuion

.BR tuxstatus
prints episode download status using the graphical
.BR ncurses
library. It can be obtained from http://github.com/kookjr/tuxstatus


CHANGELOG
Revision 227
* Fixed uninitialized variable when using the -c option of tuxcast.
--------------------------------------------------------------------------------

Revision 226
* Added option to query a named feed or a URL not subscribed to, to get
  a list of available episodes.
* Added option to download more than one episode when updating a feed.
* Fixed man page example for setting podcast directory.
--------------------------------------------------------------------------------

Revision 225
Directly use the tuxstatus graphical front end with the -s option
unless any of the following are true.
* tuxcast's stdout is not a pipe
* tuxstatus is not installed on the system in /usr/bin
* tuxcast is unable to create a pipe or fork
If any of these exceptions are true it behaves as it has in the past
and prints extra verbose output to the screen. This feature gives the
appearance that tuxcast now has an optional graphical mode from the
user's perspective.
--------------------------------------------------------------------------------

Revision 224
* Fixed two race conditions; accessing allfiles download list and
read-write of files.xml.
* Allow special exclude patterns and pattern deletion.  The special
pattern, @all@, now matches all podcast names. This can be used to
keep up to date with feed, but not actually download any
episodes. Also allow deleting patterns by adding a name that
already exists. Fixed man page for new handling of patterns.
--------------------------------------------------------------------------------

Revision 223
Fix feed parse error messages to include the feed (URL)
that failed.
--------------------------------------------------------------------------------

Revision 222
Make process lock user specific by adding user id to lockname.
--------------------------------------------------------------------------------

Revision 220, 221:
Added process lock that will keep two tuxcast/tuxcast-config's from
running at the same time (except for simple data reporting). Use
fcntl() file lock to make lock cleanup more robust.
--------------------------------------------------------------------------------

Revision 219:
Some minor makefile and packaging fixes before releasing tarball and DEB package
* po and hooks were installing in wrong directories during package build, fixed
* added tarball build for packaging
* man install/uninstall was not using prefix correctly
--------------------------------------------------------------------------------

Revision 218:
###############################################################################
===============================================================================
------------------------- Tuxcast 0.4 Released! -------------------------------
===============================================================================
###############################################################################

Revision 217:
* display feeds with empty folders better
* fix uninstall to clean up man pages and translation files
* rewrote man pages to be more useful
--------------------------------------------------------------------------------

Revision 216:
* Add enhanced status for downloading podcasts.
Add printing for podcast queuing to awaiting threads, and download
progress from curl downloads. All status prints are now thread
safe so the messages are not intermingled. Added -s option to
tuxcast to enable this printing. This can be used in combination
with tuxstatus (see: http://github.com/kookjr/tuxstatus).
--------------------------------------------------------------------------------

Revision 215:
* Allow exclusion of certain podcast episodes.
Some podcast feeds have more than one type of episode. Added the
ability to apply exclusion patterns that when matched that episode
will be treated like it is downloaded, but it will not be. Currently
the patterns are case sensitive strings. Added a -p to the -u option
of tuxcast-config to add patterns. All versions of feed display also
display the patterns.
--------------------------------------------------------------------------------

Revision 214:
* Added formatted printout for feeds in tuxcast-config -G.
Each feed is printed on one line, with the feed name, dir (id) and
feed. The columns are constant width and dynamically calculated.
The feed dir to id mapping is printed before the feeds. To minimize
the changes, I overloaded the feed::displayConfig method so other
places it is called can still have the original format.
--------------------------------------------------------------------------------

Revision 213:
* Define SUBVERSION_REVISION dynamically so you don't have to remember to change version.h each commit.
--------------------------------------------------------------------------------

Revision 212:
(And when you're compiling I'll be still alive)
* Patched cleaner.cpp to use correct output format string for the data type
	(i.e. preventing compiler warnings on newer systems.)
(Thanks to Mathew Cucuzella)
--------------------------------------------------------------------------------

Revision 211:
(It's been a while)
* Added string.h to various things.  For some reason it worked without it. Until now
* Fixed a typo in config/tuxcast-config.cpp that for some reason didn't break anything until now.
* Updated the libraries/Makefile to correctly clean bits.o
* Added the sendmail.sh hook script.  Pass it two arguments, preferably %f %p.  And have a working local mail sending system.  Stick your email address into the script.
TTFN.
(l)
(This commit dedicated to Her)
--------------------------------------------------------------------------------

Revision 210:
* Fixed a buffer overflow off-by-one error
* Set the minimum number of threads to 1
* Disabled optimisation by default
* Disabled building handle_bittorrent when torrenting is disabled
* Added the code to increment the thread counter to getlist & up2date
* Updated TODO

trunk/tuxcast/tuxcast.cpp
/*
 * 
 * This file is part of Tuxcast, "The linux podcatcher"
 * Copyright (C) 2006-2008 David Turner
 * Copyright (C) 2009, 2010 Mathew Cucuzella (kookjr@gmail.com)
 * 
 * Tuxcast is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuxcast is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Tuxcast; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */



#include "../compile_flags.h"
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <iostream>
#include <stdio.h> // :)
#include "rss.h"
#include "../config/config.h"
#include <curl/curl.h>
#include <cstdio> // Needed to open a file in the classic way, so libcurl can write to it
#include <string.h>
#include <libxml/tree.h>   // V----------------V
#include <libxml/parser.h> // for filelist stuff
#include <unistd.h>
#include "tuxcast_functions.h"
#include "tuxcast_exceptions.h"
#include "../libraries/filestuff.h"
#include "../libraries/filestuff_exceptions.h"
#include "../libraries/common.h"
#include "../config/config_exceptions.h"
#include "rss_exceptions.h"
#include "tuxcast_functions.h"
#include "cleaner.h"
#include <map>
#include "../version.h"

#ifdef THREADS
#include <pthread.h>
#endif

using namespace std;


#include <libintl.h>
#include <locale.h>


const char options[] = "scuC:U:q:fvhe:";
#define _(x) gettext(x)

static bool setup_output(void);

int main(int argc, char *argv[])
{
	configuration myconfig;
	map<char, string> vars;

        // TODO: Fix this stuff?:
	setlocale(LC_ALL,"");
	bindtextdomain("tuxcast","/usr/share/locale");
	textdomain("tuxcast");
	
	
	// Remember, we can only take one primary option.
	// We store the first option, then check - if there's another, we bork out and moan.
	
	char opt1=getopt(argc,argv,options);
	string optarg1;
	if(optarg != NULL)
		optarg1=optarg;
	else
		optarg1="";
	
        // -u and -U may have second option of -e num
	char opt2=getopt(argc,argv,options);
        int episode_cnt = 1;
        if (opt2 != -1 && opt2 != 'e')
	{
		cerr << "Error, more than one option was passed and not -u or -U.  You must only pass one option." << endl;
		return -1;
	}
        else if (opt2 == 'e' && (episode_cnt = atoi(optarg)) <= 0)
        {
		cerr << "Error, invalid episode count for -e." << endl;
		return -1;
        }

	if(!((opt1 == 'h') || (opt1 == 'v'))) // Load config if option isn't h or v
	{
		try
		{
			myconfig.load();
		}
		catch(eConfig_NoConfigFile &e)
		{
			fprintf(stderr,_("Cannot load config file - please create one\n"));
			return -1; // No need to print the exception
			// We know exactly what this one means
		}
		// If theres no config file, then
		// a) checkfileexists will throw an exception, which will be
		// caught here, and we abort.
		// b) checkfileexists will return false
	}

	try
	{
                bool subprocess = false;
		switch(opt1)
		{
                        case 's':
			case 'c':
                                set_lock();
                                if (opt1 == 's') {
                                    myconfig.progress = true;
                                    subprocess = setup_output();
                                }
				printf(_("Checking all feeds\n"));
				checkall(myconfig);
				setvars(vars, myconfig);
				runhook(POSTRUN, vars, myconfig);
				xmlCleanupParser();
                                if (subprocess) {
                                    fclose(stdout);  /* tuxstatus will see EOF */
                                    wait(NULL);      /* wait for child */
                                }
				break;
			case 'u':
                                set_lock();
				printf(_("Getting up to date on all feeds\n"));
				up2dateall(myconfig, episode_cnt);
				setvars(vars, myconfig);
				runhook(POSTRUN, vars, myconfig);
				xmlCleanupParser();
				break;

			case 'C':
                                set_lock();
				printf(_("Checking feed, \"%s\"\n"), optarg1.c_str());
				// We need to loop through myconfig.feeds to find the feed ID corresponding to the passed name
				if(strcmp(optarg1.c_str(),"") == 0)
				{
					fprintf(stderr,_("You must pass a non-blank feed name\n"));
					return -1;
				}
				FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed)
				{
					if(strcasecmp(optarg1.c_str(),feed->name.c_str()) == 0)
					{
						// Found the feed
						filelist files;
						check(myconfig, *feed, files);
						getlist(files, myconfig);
#ifdef THREADS
						for(int i=0; i<myconfig.threads.size(); i++)
							pthread_join(myconfig.threads[i], NULL);
#endif
						setvars(vars, myconfig);
						runhook(POSTRUN, vars, myconfig);
						xmlCleanupParser();
						return 0;
					}
				}
				// If we got all through the feeds, and it wasn't found (and we returned),
				// then the feed doesn't exist:
				fprintf(stderr,_("Unknown feed, \"%s\"\n"),optarg1.c_str());
				return -1;
				break; // Bah
			case 'U':
                                set_lock();
				printf(_("Getting up2date on feed \"%s\"\n"),optarg1.c_str());
				// We need to loop through myconfig.feeds to find the feed ID corresponding to the passed name
				if(strcmp(optarg1.c_str(),"") == 0)
				{
					fprintf(stderr,_("You must pass a non-blank feed name\n"));
					return -1;
				}
				FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed)
				{
					if(strcasecmp(optarg1.c_str(),feed->name.c_str()) == 0)
					{
						// Found the feed
						filelist files;
						up2date(myconfig, *feed, files, episode_cnt);
						getlist(files, myconfig);
#ifdef THREADS
						for(int i=0; i<myconfig.threads.size(); i++)
							pthread_join(myconfig.threads[i], NULL);
#endif
						setvars(vars, myconfig);
						runhook(POSTRUN, vars, myconfig);
						xmlCleanupParser();
						return 0;
					}
				}
				// If we got all through the feeds, and it wasn't found (and we returned),
				// then the feed doesn't exist:
				fprintf(stderr,_("Unknown feed, \"%s\"\n"),optarg1.c_str());
				return -1;
				break; // Bah

			case 'q':
                                set_lock();
				cout << "Querying feed for available episodes..." << endl;

                                if (optarg1.find("http://") == 0 || optarg1.find("https://") == 0) {
                                    show_episodes(myconfig, optarg1);
                                }
                                else {
                                    if (optarg1.size() == 0) {
					fprintf(stderr,_("You must pass a non-blank feed name\n"));
					return -1;
                                    }
                                    FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed) {
					if(strcasecmp(optarg1.c_str(),feed->name.c_str()) == 0) {
                                            show_episodes(myconfig, feed->address);
                                        }
                                    }
                                }
				xmlCleanupParser();
				break;
                                
			case 'f':
                                set_lock();
				cout << "Cleaning out files.xml..." << endl;
				clean();
				break;

			case 'v':
                                // no lock checking needed here, just info
				version();
				break;

			case 'h':
                                // no lock checking needed here, just info
				// Fallthrough:

			default:
                                // no lock checking needed here, just info
				printf(_("Usage: tuxcast <option>\n"));
				printf(_("where <option> is one of the below:\n"));
				printf(_("-s - Check all feeds, verbose output\n"));
				printf(_("-c - Check all feeds\n"));
				printf(_("-u [-e num] - Download only the latest num files from all feeds (default 1)\n"));
				printf(_("-C name - Download all episodes of the named feed\n"));
				printf(_("-U name [-e num] - Download only the latest episode of the named feed (default 1)\n"));
				printf(_("-q name|URL - Query feed for available episodes\n"));
				printf(_("-h - Show this help message\n"));
				printf(_("-f - clean out old podcasts from files.xml\n"));
				printf(_("-v - Show version and license information\n"));
		}
	}
	catch(eTuxcast_FSFull &e)
	{
		fprintf(stderr, _("Error: could not save config/state due to full filesystem:\n"));
		e.print();

		return -1;
	}
	catch(eProcessLock &e)
	{
		e.print();

		return -1;
	}

	return 0;
}

/*
 * If possible redirect the standard output to child process of
 * tuxstatus through a pipe. Return true if success and the
 * child process is spawned.
 */
static bool setup_output(void) {
    int pipefd[2];
    pid_t cpid;
    struct stat st;
    const char* tuxstatus_prog = "/usr/bin/tuxstatus";

    /*
     * If stdout is already a pipe, assume the user has already
     * done something line, "tuxcast -s | tuxstatus" and abort.
     */
    if (fstat(STDOUT_FILENO, &st) == -1 || (st.st_mode & S_IFIFO)) {
        return false;
    }

    /*
     * Make sure the program tuxstatus is available to exec
     */
    if (stat(tuxstatus_prog, &st) == -1) {
        return false;
    }

    /*
     * Create a pipe that will be used for tuxcast's stdout to be
     * hooked to tuxstatus' stdin.
     */
    if (pipe(pipefd) == -1) {
        return false;
    }

    /*
     * Fork the process. The child will become tuxstatus.
     */
    if ( (cpid=fork()) == -1 ) {
        return false;
    }

    if (cpid == 0) {   /* this is the child process, setup pipe as stdin */

        close(pipefd[1]);  /* close unused write end */

        /* redirect stdin to the pipe */
        FILE* tmp = fdopen(pipefd[0], "r");
        dup2(tmp->_fileno, STDIN_FILENO);
        fclose(tmp);

        /* exec the tuxstatus program */
        execl(tuxstatus_prog, "tuxstatus", "-d", (char* )NULL);

        /* only get here if exec fails */
        _exit(EXIT_SUCCESS);

    }
    else {    /* parent */

        close(pipefd[0]);   /* close unused read end */

        /* redirect stdout to pipe */
        FILE* tmp = fdopen(pipefd[1], "w");
        dup2(tmp->_fileno, STDOUT_FILENO);
        fclose(tmp);

        /* keep error messages from messing up ncurses formatted output */
        fclose(stderr);

        return true;
    }
}
